pkg/operator/ceph/cluster/controller.go

```diff
@@ -214,64 +216,182 @@ func (c *ClusterController) onK8sNodeAdd(obj interface{}) {
 }
 
 func (c *ClusterController) onAdd(obj interface{}) {
-	// notify the callback that a cluster crd is being added
-	if c.addClusterCallback != nil {
-		if err := c.addClusterCallback(); err != nil {
-			logger.Errorf("%+v", err)
-		}
-	}
 
 	clusterObj, err := getClusterObject(obj)
 	if err != nil {
 		logger.Errorf("failed to get cluster object: %+v", err)
 		return
 	}
 
+	if existing, ok := c.clusterMap[clusterObj.Namespace]; ok {
+		logger.Errorf("Failed to add cluster cr %s in namespace %s. Cluster cr %s already exists in this namespace. Only one cluster cr per namespace is supported.",
+			clusterObj.Name, clusterObj.Namespace, existing.crdName)
+		return
+	}
+
 	cluster := newCluster(clusterObj, c.context)
 	c.clusterMap[cluster.Namespace] = cluster
 
 	logger.Infof("starting cluster in namespace %s", cluster.Namespace)
 
+	// notify the callback that a cluster crd is being added
+	if c.addClusterCallback != nil {
+		if err := c.addClusterCallback(cluster.Spec.External.Enable); err != nil {
+			logger.Errorf("%+v", err)
+		}
+	}
+
+	if !cluster.Spec.External.Enable {
+		if err := c.configureLocalCephCluster(clusterObj.Namespace, clusterObj.Name, cluster); err != nil {
+			logger.Errorf("failed to configure local ceph cluster. %+v", err)
+			return
+		}
+	} else {
+		if err := c.configureExternalCephCluster(clusterObj.Namespace, clusterObj.Name, cluster); err != nil {
+			logger.Errorf("failed to configure external ceph cluster. %+v", err)
+			return
+		}
+	}
+
+	// Start pool CRD watcher
+	poolController := pool.NewPoolController(c.context, cluster.Spec)
+	poolController.StartWatch(cluster.Namespace, cluster.stopCh)
+
+	// Start object store CRD watcher
+	objectStoreController := object.NewObjectStoreController(cluster.Info, c.context, cluster.Namespace, c.rookImage, cluster.Spec, cluster.ownerRef, cluster.Spec.DataDirHostPath)
+	objectStoreController.StartWatch(cluster.Namespace, cluster.stopCh)
+
+	// Start file system CRD watcher
+	fileController := file.NewFilesystemController(cluster.Info, c.context, cluster.Namespace, c.rookImage, cluster.Spec, cluster.ownerRef, cluster.Spec.DataDirHostPath)
+	fileController.StartWatch(cluster.Namespace, cluster.stopCh)
+
+	// Start nfs ganesha CRD watcher
+	ganeshaController := nfs.NewCephNFSController(cluster.Info, c.context, cluster.Spec.DataDirHostPath, cluster.Namespace, c.rookImage, cluster.Spec, cluster.ownerRef)
+	ganeshaController.StartWatch(cluster.Namespace, cluster.stopCh)
+
+	// Populate ClusterInfo
+	if cluster.Spec.External.Enable {
+		cluster.mons.ClusterInfo = cluster.Info
+	}
+
+	// Start mon health checker
+	healthChecker := mon.NewHealthChecker(cluster.mons, cluster.Spec)
+	go healthChecker.Check(cluster.stopCh)
```

Comment:
The mon health checker is where we will see if the mons changed in the external cluster?

Reply:
Yes, with the help of aec7358bf7260ea52a9e149e4b527fa39582c074